File: lib/fastmail.rb
class FastmailConfig
  attr_accessor :name, :redirectTo, :fileIn, :search, :markRead,
    :showNotification, :skipInbox, :discard, :markSpam, :markFlagged,
    :conditions, :snoozeUntil, :previousFileInName, :combinator

  DEFAULT = {
    name: "",
    combinator: "any",
    markRead: false,
    stop: false,
    fileIn: nil,
    redirectTo: nil,
    skipInbox: false,
    showNotification: false,

    # Don't really know anything other than how we're sorting
    discard: false,
    markSpam: false,
    markFlagged: false,

    # Features we're not using
    conditions: nil,
    snoozeUntil: nil,
    previousFileInName: nil,
  }

  def initialize(values)
    values.each do |key, value|
      self.send("#{key}=", value)
    end
  end

  def to_rule
    Rule.new(
      name: self.name,
      redirect: self.redirectTo[0],
      folder: self.fileIn,
      match: parse_search(self.search),
      mark_read: self.markRead,
      notify: self.showNotification,
      archive: self.skipInbox,
    )
  end

  def self.from_rule(rule)
    FastmailConfig.new(
      name: rule.name,
      redirectTo: ([rule.redirect] if rule.redirect),
      fileIn: rule.folder,
      search: generate_search(rule.match),
      markRead: rule.mark_read,
      showNotification: rule.notify,
      skipInbox: rule.archive,
    )
  end

  def to_h
    DEFAULT.dup.merge({
      name: self.name,
      redirectTo: self.redirectTo,
      fileIn: self.fileIn,
      search: self.search,
      markRead: self.markRead,
      showNotification: self.showNotification,
      skipInbox: self.skipInbox,
      discard: self.discard,
      markSpam: self.markSpam,
      markFlagged: self.markFlagged,
      conditions: self.conditions,
      snoozeUntil: self.snoozeUntil,
      previousFileInName: self.previousFileInName,
      combinator: self.combinator,
    })
  end
end

def generate_search(params)
  params.map do |field, values|
    values.map do |value|
      if value.include?(" ")
        value = "(#{value})"
      end
      "#{field}:#{value}"
    end
  end.flatten.join(" OR ")
end

def parse_search(query)
  search = Hash.new([])
  query.split(" OR ").map do |term|
    field, value = term.split(":")
    value = value.sub(/^\(/, "").sub(/\)$/, "")
    search[field] += [value]
    search[field].sort!
  end
  search
end

File: rules/fastmail.py
"""
Model for understanding Fastmail config files.

    mailrules.json

"""

import json
from datetime import datetime
from typing import Union, List
from pydantic import BaseModel, Field

from rules.common import CommonRule, CommonMatch
from rules.ruleset import Ruleset


class FastmailRule(BaseModel):
    """An individual Fastmail mail routing rule."""

    # What's this rule about?
    name: str

    # What are we looking for in the message?
    combinator: str
    search: str
    conditions: Union[str, None]
    previousFileInName: Union[str, None]

    # Where should this message go?
    fileIn: Union[str, None]  # label or folder
    discard: bool  # delete
    redirectTo: Union[List[str], None]
    snoozeUntil: Union[dict, None]

    # What should we do with this message?
    markFlagged: bool
    markRead: bool
    markSpam: bool
    showNotification: bool
    skipInbox: bool  # archive

    # Can we do anything else after this rule?
    stop: bool

    created: datetime = Field(default_factory=datetime.now)
    updated: datetime = Field(default_factory=datetime.now)

    @classmethod
    def from_common_rule(cls, rule: CommonRule) -> "FastmailRule":
        """Convert a CommonRule to a FastmailRule."""

        return cls(
            name=rule.name,
            combinator="any",
            search=generate_search(rule.match),
        )


class FastmailRuleset(Ruleset):
    """A collection of Fastmail rules."""

    rules: List[FastmailRule]
    rule_class = FastmailRule
    file_loader = json
    default_filename = "mailrules.json"

    suffix = ".json"

    @classmethod
    def from_common_ruleset(cls, ruleset: "CommonRuleset") -> "FastmailRuleset":
        """Convert a CommonRuleset to a FastmailRuleset."""

        return cls(
            rules=[FastmailRule.from_common_rule(rule) for rule in ruleset.rules]
        )


def generate_search(match: CommonMatch) -> str:
    """Convert a match into a search query"""

    query = ""
    return query
    # for (key, values) in match.dict().items():


# def generate_search(params)
#   params.map do |field, values|
#     values.map do |value|
#       if value.include?(" ")
#         value = "(#{value})"
#       end
#       "#{field}:#{value}"
#     end
#   end.flatten.join(" OR ")
# end
